test/src/test/java/org/corfudb/integration/MergeTestListener.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    @Before
+    public void init() throws IOException, InvocationTargetException, NoSuchMethodException, IllegalAccessException {
+        new AbstractIT.CorfuServerRunner()
+                .setHost(corfuSingleNodeHost)
+                .setPort(corfuStringNodePort)
+                .setLogPath(getCorfuServerLogPath(corfuSingleNodeHost, corfuStringNodePort))
+                .setSingle(true)
+                .runServer();
+        runtime = createRuntime(singleNodeEndpoint);
+        store = new CorfuStore(runtime);
+        openMergedTable();
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        log.error("Error on listener", throwable);
+        fail("onError for LRTestListener: " + throwable.toString());
+    }
+
+    /**
+     * Invoke when snapshot sync starts
+     */
+    @Override
+    protected void onSnapshotSyncStart() {
+        log.info("Snapshot sync started");
+    }
+
+    /**
+     * Apply addedKeys and deletedKeys to merged table, add keys to
+     * existingKeysInMergedTable from addedKeys and deletedKeys.
+     */
+    @Override
+    protected void onSnapshotSyncComplete() {
+        existingKeysInMergedTable.addAll(addedKeys.keySet());
+        addedKeys.forEach((addedKey, e) -> {
+            try (TxnContext tx = store.txn(namespace)){
+                tx.putRecord(mergedTable, (SampleSchema.Uuid) addedKey,
+                        (SampleSchema.ValueFieldTagOne) e.getPayload(), (SampleSchema.Uuid) e.getMetadata());
+                tx.commit();
+            }
+        });
+
+        existingKeysInMergedTable.addAll(deletedKeys);
+        deletedKeys.forEach(deletedKey -> {
+            try (TxnContext tx = store.txn(namespace)){
+                tx.delete(mergedTable, (SampleSchema.Uuid) deletedKey);
+                tx.commit();
+            }
+        });
+        log.info("Snapshot sync complete");
+    }
+
+    /**
+     * Write local tables to merged table, process updates and add the keys to addedKeys
+     * and deletedKeys for replicated table.
+     *
+     * @param results Entries received in a single transaction as part of a snapshot sync
+     */
+    @Override
+    protected void processUpdatesInSnapshotSync(CorfuStreamEntries results) {
+        log.info("Processing updates in snapshot sync");
+        results.getEntries().forEach((schema, entries) -> {
+            entries.forEach(e -> {
+                Message key = e.getKey();
+                if (schema.getTableName().equals("local_table_name")) {
+                    existingKeysInMergedTable.add(key);
+                    try (TxnContext tx = store.txn(namespace)){
+                        tx.putRecord(mergedTable, (SampleSchema.Uuid) key, (SampleSchema.ValueFieldTagOne)
+                                        e.getPayload(),
+                                (SampleSchema.Uuid) e.getMetadata());
+                        tx.commit();
+                    }
+                }
```

Comment:
https://confluence.eng.vmware.com/pages/viewpage.action?spaceKey=NSBU&title=Log+Replicator+Design+for+SaaS
I thought we only need to update the existingKeysInMergedTable when the listener is processing replicated tables since the comment "// Update deletedKeys and addedKeys using the operation type by performing a diff against existingKeysInMergedTable" is in the else block of processUpdatesInSnapshotSync(). 

Will update this.